### PR TITLE
[ewasm] Clean up some builtins and fix text representation of `select`

### PR DESCRIPTION
**Pull request by axic**
Fixes #10430.

### CODE REVIEW

File: libyul/backends/wasm/BinaryTransform.cpp [hunk 1]
```diff
@@ -381,30 +376,21 @@ bytes BinaryTransform::operator()(BuiltinCall const& _call)
 		return toBytes(Opcode::I64Const) + lebEncodeSigned(static_cast<int64_t>(m_subModulePosAndSize.at(name).second));
 	}
 
+	yulAssert(builtins.count(_call.functionName), "Builtin " + _call.functionName + " not found");
+	// NOTE: the dialect ensures we have the right amount of arguments
 	bytes args = visit(_call.arguments);
+	bytes ret = move(args) + toBytes(builtins.at(_call.functionName));
```
Comment (axic) [inline libyul/backends/wasm/BinaryTransform.cpp:382]: Should this be ?
```suggestion
	bytes ret = visit(_call.arguments) + toBytes(builtins.at(_call.functionName));
```

File: libyul/backends/wasm/BinaryTransform.cpp [hunk 2]
```diff
@@ -381,30 +376,21 @@ bytes BinaryTransform::operator()(BuiltinCall const& _call)
...
+		// actually be more expensive. It's best to hint at 1-byte alignment if we don't plan
+		// to control the memory layout accordingly.
+		ret += bytes{{0, 0}}; // 2^0 == 1-byte alignment
 
-	if (_call.functionName == "unreachable")
-		return toBytes(Opcode::Unreachable);
-	else if (_call.functionName == "nop")
-		return toBytes(Opcode::Nop);
-	else if (_call.functionName == "i32.drop" || _call.functionName == "i64.drop")
-		return toBytes(Opcode::Drop);
```
Comment (axic) [inline libyul/backends/wasm/BinaryTransform.cpp:391]: I think having drop here is actually a bug. The arguments were visited, but their results were not stored in the bytecode. This meant we were removing bytecode which could have side-effects.

Hence made it part of the regular code path. If we want to add optimisations steps for drop, that should be separate.

File: libyul/backends/wasm/BinaryTransform.cpp (hunk 1 above)
Comment (chriseth) [reply libyul/backends/wasm/BinaryTransform.cpp:382]: I don't think it makes a difference, but having a call to `visit` stand out is probably a good thing - also concerning the order of evaluation.

File: test/libyul/objectCompiler/wasm/no_main_function.yul [hunk 3]
```diff
@@ -19,4 +19,4 @@
 // )
 //
 // Binary:
-// 0061736d01000000010401600000020100030201000503010001060100070a01066d656d6f727902000a0801060002401a0b0b
```
Comment (chriseth) [inline test/libyul/objectCompiler/wasm/no_main_function.yul:22]: Why did it get longer?

**Review by chriseth — APPROVED** [review]

File: libyul/backends/wasm/BinaryTransform.cpp (hunk 1 above)
Comment (axic) [reply libyul/backends/wasm/BinaryTransform.cpp:382]: > also concerning the order of evaluation.

That's why I left it this way, but a lot of other places in the code actually have it in a single statement.

File: test/libyul/objectCompiler/wasm/no_main_function.yul (hunk 3 above)
Comment (axic) [reply test/libyul/objectCompiler/wasm/no_main_function.yul:22]: https://github.com/ethereum/solidity/pull/10454#discussion_r533383239

It was not including the arguments to drop. Which in the above cases were side-effect free, but nothing ensured that.

**Pull request merged**